src/nvim/msgpack_rpc/channel.c

```diff
@@ -500,7 +500,12 @@ static void rpc_close_event(void **argv)
       // Avoid hanging when there are no other UIs and a prompt is triggered on exit.
       remote_ui_disconnect(channel->id);
     }
-
+    if (is_ui_client && ui_client_pending_restart) {
+      ui_client_do_restart();
+      if (ui_client_channel_id != channel->id) {  // restart success
+        return;
```

Comment:
should we ELOG a message on failure ? I don't remember if the UI client can log things though, but maybe it can emit something to the server (though that might not even exist), or maybe to stderr?

Reply:
I've renamed `ui_client_do_restart()` to `ui_client_may_restart_server()` and moved the pending check into that function.